core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -846,17 +852,16 @@ class ReplicaManager(val config: KafkaConfig,
                 s"Unable to verify the partition has been added to the transaction. Underlying error: ${error.toString}"))
               case _ => None
             }
-          topicPartition -> LogAppendResult(
+          new TopicIdPartition(topicIds.getOrElse(topicPartition.topic(), Uuid.ZERO_UUID), topicPartition) -> LogAppendResult(
```

Comment:
hmm, in theory yes this might happened. However `topicIds` is build based on `entriesPerPartition.keys` so it is very unlikely that we reach this point with unmatched as far as I can see. However for safety I have updated the filter of `entriesPerPartition` that form `entriesWithoutErrorsPerPartition` to relay on topicPartition.